File: src/server/src/test/test_velvet_dawn/test_mechanics/test_conditionals/test_conditional_count.py
import velvet_dawn
from test.base_test import BaseTest
from velvet_dawn import errors
from velvet_dawn.dao import app
from velvet_dawn.models.coordinate import Coordinate


class TestConditionalCount(BaseTest):

    def test_conditional_count(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.db.units.get_units_at_positions(Coordinate(5, 0))[0]

            conditional_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "entities", "equals": 4})
            conditional_not_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "entities", "ne": 4})
            conditional_less_than = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "lt": 4})
            conditional_less_than_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "lte": 4})
            conditional_greater_than = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "gt": 4})
            conditional_greater_than_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "gte": 4})

            # cant compare tag on attribute
            with self.assertRaises(errors.ValidationError):
                velvet_dawn.mechanics.conditionals.get_conditional(
                    "0", {"count": "self", "tagged": "tag:x"})

            self.assertTrue(conditional_equals.is_true(unit))
            self.assertFalse(conditional_not_equals.is_true(unit))
            self.assertFalse(conditional_less_than.is_true(unit))
            self.assertTrue(conditional_less_than_equals.is_true(unit))
            self.assertFalse(conditional_greater_than.is_true(unit))
            self.assertTrue(conditional_greater_than_equals.is_true(unit))

    def test_conditional_if_function_value(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.units.list()[0]
            unit.set_attribute("units-count-true", 4)
            unit.set_attribute("units-count-false", 5)

            conditional_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "equals": "@self.units-count-true"})
            conditional_not_equals = velvet_dawn.mechanics.conditionals.get_conditional(
                "0", {"count": "units", "equals": "@self.units-count-false"})

            self.assertTrue(conditional_equals.is_true(unit))
            self.assertFalse(conditional_not_equals.is_true(unit))

File: src/jserver/src/test/velvetdawn/mechanics/conditionals/TestConditionalCount.java
package test.velvetdawn.mechanics.conditionals;

import org.junit.Test;
import test.BaseTest;
import velvetdawn.mechanics.conditionals.Conditionals;
import velvetdawn.models.Coordinate;
import velvetdawn.models.anytype.Any;
import velvetdawn.utils.Json;

import java.util.ArrayList;

import static junit.framework.TestCase.assertFalse;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;

public class TestConditionalCount extends BaseTest {

    @Test
    public void test_conditional_count() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);

        var conditionalEquals = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("equals", 4));
        var conditionalNotEquals = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("ne", 4));
        var conditionalLessThan = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("lt", 4));
        var conditionalLessThanEquals = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("lte", 4));
        var conditionalGreaterThan = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("gt", 4));
        var conditionalGreaterThanEquals = Conditionals.get(velvetDawn, "0", new Json()
                .set("count", "entities")
                .set("gte", 4));

        // can't compare tag on attribute
        assertThrows(Exception.class, () -> Conditionals.get(velvetDawn, "", new Json()
                .set("count", "self")
                .set("tagged", "atag")));

        assertTrue(conditionalEquals.isTrue(unit));
        assertFalse(conditionalNotEquals.isTrue(unit));
        assertFalse(conditionalLessThan.isTrue(unit));
        assertTrue(conditionalLessThanEquals.isTrue(unit));
        assertFalse(conditionalGreaterThan.isTrue(unit));
        assertTrue(conditionalGreaterThanEquals.isTrue(unit));
    }

    @Test
    public void test_conditional_if_function_value() throws Exception {
        var velvetDawn = this.prepareGame();

        var unit = new ArrayList<>(velvetDawn.entities.list()).get(0);
        unit.attributes.set("units-count-true", Any.from(4));
        unit.attributes.set("units-count-false", Any.from(5));

        var conditionalEquals = Conditionals.get(velvetDawn, "", new Json()
                .set("count", "entities")
                .set("equals", "@self.units-count-true"));
        var conditionalNotEquals = Conditionals.get(velvetDawn, "", new Json()
                .set("count", "entities")
                .set("equals", "@self.units-count-false"));

        assertTrue(conditionalEquals.isTrue(unit));
        assertFalse(conditionalNotEquals.isTrue(unit));
    }
}
